numpy/ma/core.py

```diff
@@ -3171,7 +3164,7 @@ def __getitem__(self, indx):
                 dout._mask = _mask[indx]
                 # set shape to match that of data; this is needed for matrices
                 dout._mask.shape = dout.shape
-                dout._sharedmask = True
+                dout._sharedmask = np.may_share_memory(_mask, dout._mask)
```

Comment:
Actually, I guess I could undo this, in favor of doing that slightly silly refcount + owndata check below, which should find it. I think I will do that.